msgio.go

```diff
@@ -152,16 +158,27 @@ func NewReaderWithPool(r io.Reader, p *mpool.Pool) ReadCloser {
 // NextMsgLen reads the length of the next msg into s.lbuf, and returns it.
 // WARNING: like Read, NextMsgLen is destructive. It reads from the internal
 // reader.
-func (s *reader) NextMsgLen() (int, error) {
+func (s *reader) NextMsgLen() (uint32, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 	return s.nextMsgLen()
 }
 
-func (s *reader) nextMsgLen() (n int, err error) {
-	if s.next == -1 {
-		s.next, err = ReadLen(s.R, s.lbuf)
+func (s *reader) nextMsgLen() (uint32, error) {
+	var err error
+	if !s.more {
+		return 0, ErrBadCall
+	}
+
+	s.next, err = ReadLen(s.R, s.lbuf)
+	if err != nil {
+		return 0, err
 	}
+
+	if s.next > s.max {
+		return 0, ErrMsgTooLarge
+	}
+
```

Comment:
look at the older code. it's meant to be able to call `nextMsgLen` multiple times.